File: rust/src/world.rs
use crate::hit_record::HitRecord;
use crate::ray::Ray;
use crate::sphere::Sphere;

pub struct World {
    pub objects: Vec<Sphere>,
}

impl World {
    pub fn hit(&self, ray: &Ray, t_min: f32, t_max: f32) -> Option<HitRecord> {
        let mut closest = t_max;
        let mut result = None;
        for h in self.objects.iter() {
            let hr = h.hit(ray, t_min, closest);
            if hr.is_some() {
                result = hr;
                closest = result.as_ref().unwrap().t;
            }
        }
        result
    }
}

File: nim/src/world.nim
import std/options

import sphere
import ray
import material

type
  World* = object
    objects: seq[Sphere]

func hit*(self: World; ray: Ray; t_min, t_max: float): Option[HitRecord] =
  var closest = t_max
  for sphere in self.objects:
    let hr = sphere.hit(ray, t_min, closest)
    if hr.isSome():
      result = hr
      closest = hr.get().t

func add*(self: var World; s: Sphere) =
  self.objects.add(s)
